File: core-android/src/main/java/com/lookaround/core/android/view/BoxedVerticalSeekbar.java
package com.lookaround.core.android.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Region;
import android.graphics.drawable.BitmapDrawable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.core.content.ContextCompat;

import com.lookaround.core.android.R;

public class BoxedVerticalSeekbar extends View {
    private static final int MAX = 100;
    private static final int MIN = 0;

    /**
     * The min value of progress value.
     */
    private int min = MIN;

    /**
     * The Maximum value that this SeekArc can be set to
     */
    private int max = MAX;

    /**
     * The increment/decrement value for each movement of progress.
     */
    private int step = 10;

    /**
     * The corner radius of the view.
     */
    private int cornerRadius = 10;

    /**
     * Text size in SP.
     */
    private float textSize = 26;

    /**
     * Text bottom padding in pixel.
     */
    private int textBottomPadding = 20;

    private int points;

    private boolean enabled = true;
    /**
     * Enable or disable text .
     */
    private boolean textEnabled = true;

    /**
     * Enable or disable image .
     */
    private boolean imageEnabled = false;

    private boolean snapEnabled = true;

    /**
     * mTouchDisabled touches will not move the slider only swipe motion will activate it
     */
    private boolean touchDisabled = true;

    private float progressSweep = 0;
    private Paint progressPaint;
    private Paint textPaint;
    private int scrWidth;
    private int scrHeight;
    private OnValuesChangeListener onValuesChangeListener;
    private int backgroundColor;
    private int defaultValue;
    private Bitmap defaultImage;
    private Bitmap minImage;
    private Bitmap maxImage;
    private final Rect rect = new Rect();
    private boolean firstRun = true;
    private String pointsText = String.valueOf(defaultValue);

    public BoxedVerticalSeekbar(Context context) {
        super(context);
        init(context, null);
    }

    public BoxedVerticalSeekbar(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        System.out.println("INIT");
        float density = getResources().getDisplayMetrics().density;

        // Defaults, may need to link this into theme settings
        int progressColor = ContextCompat.getColor(context, R.color.color_progress);
        backgroundColor = ContextCompat.getColor(context, R.color.color_background);
        backgroundColor = ContextCompat.getColor(context, R.color.color_background);

        int textColor = ContextCompat.getColor(context, R.color.color_text);
        textSize = (int) (textSize * density);
        defaultValue = max / 2;

        if (attrs != null) {
            final TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.BoxedVerticalSeekbar, 0, 0);

            points = a.getInteger(R.styleable.BoxedVerticalSeekbar_points, points);
            max = a.getInteger(R.styleable.BoxedVerticalSeekbar_max, max);
            min = a.getInteger(R.styleable.BoxedVerticalSeekbar_min, min);
            step = a.getInteger(R.styleable.BoxedVerticalSeekbar_step, step);
            defaultValue = a.getInteger(R.styleable.BoxedVerticalSeekbar_defaultValue, defaultValue);
            cornerRadius = a.getInteger(R.styleable.BoxedVerticalSeekbar_libCornerRadius, cornerRadius);
            textBottomPadding =
                    a.getInteger(R.styleable.BoxedVerticalSeekbar_textBottomPadding, textBottomPadding);
            // Images
            imageEnabled = a.getBoolean(R.styleable.BoxedVerticalSeekbar_imageEnabled, imageEnabled);

            if (imageEnabled) {
                if (a.getDrawable(R.styleable.BoxedVerticalSeekbar_defaultImage) == null) {
                    throw new IllegalArgumentException(
                            "When images are enabled, defaultImage can not be null. Please assign a drawable in the layout XML file");
                }
                if (a.getDrawable(R.styleable.BoxedVerticalSeekbar_minImage) == null) {
                    throw new IllegalArgumentException(
                            "When images are enabled, minImage can not be null. Please assign a drawable in the layout XML file");
                }
                if (a.getDrawable(R.styleable.BoxedVerticalSeekbar_maxImage) == null) {
                    throw new IllegalArgumentException(
                            "When images are enabled, maxImage can not be null. Please assign a drawable in the layout XML file");
                }
                defaultImage =
                        ((BitmapDrawable) a.getDrawable(R.styleable.BoxedVerticalSeekbar_defaultImage)).getBitmap();
                minImage = ((BitmapDrawable) a.getDrawable(R.styleable.BoxedVerticalSeekbar_minImage)).getBitmap();
                maxImage = ((BitmapDrawable) a.getDrawable(R.styleable.BoxedVerticalSeekbar_maxImage)).getBitmap();
            }

            progressColor = a.getColor(R.styleable.BoxedVerticalSeekbar_progressColor, progressColor);
            backgroundColor = a.getColor(R.styleable.BoxedVerticalSeekbar_backgroundColor, backgroundColor);

            textSize = (int) a.getDimension(R.styleable.BoxedVerticalSeekbar_textSize, textSize);
            textColor = a.getColor(R.styleable.BoxedVerticalSeekbar_textColor, textColor);

            enabled = a.getBoolean(R.styleable.BoxedVerticalSeekbar_enabled, enabled);
            touchDisabled = a.getBoolean(R.styleable.BoxedVerticalSeekbar_touchDisabled, touchDisabled);
            textEnabled = a.getBoolean(R.styleable.BoxedVerticalSeekbar_textEnabled, textEnabled);
            snapEnabled = a.getBoolean(R.styleable.BoxedVerticalSeekbar_snapEnabled, snapEnabled);

            points = defaultValue;

            a.recycle();
        }

        // range check
        points = Math.min(points, max);
        points = Math.max(points, min);

        progressPaint = new Paint();
        progressPaint.setColor(progressColor);
        progressPaint.setAntiAlias(true);
        progressPaint.setStyle(Paint.Style.STROKE);

        textPaint = new Paint();
        textPaint.setColor(textColor);
        textPaint.setAntiAlias(true);
        textPaint.setStyle(Paint.Style.FILL);
        textPaint.setTextSize(textSize);

        scrHeight = context.getResources().getDisplayMetrics().heightPixels;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        scrWidth = getDefaultSize(getSuggestedMinimumWidth(), widthMeasureSpec);
        scrHeight = getDefaultSize(getSuggestedMinimumHeight(), heightMeasureSpec);
        progressPaint.setStrokeWidth(scrWidth);

        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        Paint paint = new Paint();

        paint.setAlpha(255);
        canvas.translate(0, 0);
        Path path = new Path();
        path.addRoundRect(
                new RectF(0, 0, scrWidth, scrHeight), cornerRadius, cornerRadius, Path.Direction.CCW);
        canvas.clipPath(path, Region.Op.INTERSECT);
        paint.setColor(backgroundColor);
        paint.setAntiAlias(true);
        canvas.drawRect(0, 0, scrWidth, scrHeight, paint);

        canvas.drawLine(
                canvas.getWidth() / 2,
                canvas.getHeight(),
                canvas.getWidth() / 2,
                progressSweep,
                progressPaint);

        if (imageEnabled && defaultImage != null && minImage != null && maxImage != null) {
            // If image is enabled, text will not be shown
            if (points == max) {
                drawIcon(maxImage, canvas);
            } else if (points == min) {
                drawIcon(minImage, canvas);
            } else {
                drawIcon(defaultImage, canvas);
            }
        } else {
            // If image is disabled and text is enabled show text
            if (textEnabled) {
                drawText(canvas, textPaint, pointsText);
            }
        }

        if (firstRun) {
            firstRun = false;
            setValue(points);
        }
    }

    private void drawText(Canvas canvas, Paint paint, String text) {
        canvas.getClipBounds(rect);
        int cWidth = rect.width();
        paint.setTextAlign(Paint.Align.LEFT);
        paint.getTextBounds(text, 0, text.length(), rect);
        float x = cWidth / 2f - rect.width() / 2f - rect.left;
        canvas.drawText(text, x, canvas.getHeight() - textBottomPadding, paint);
    }

    private void drawIcon(Bitmap bitmap, Canvas canvas) {
        bitmap = getResizedBitmap(bitmap, canvas.getWidth() / 2, canvas.getWidth() / 2);
        canvas.drawBitmap(
                bitmap,
                null,
                new RectF(
                        (canvas.getWidth() / 2) - (bitmap.getWidth() / 2),
                        canvas.getHeight() - bitmap.getHeight(),
                        (canvas.getWidth() / 3) + bitmap.getWidth(),
                        canvas.getHeight()),
                null);
    }

    private Bitmap getResizedBitmap(Bitmap bm, int newHeight, int newWidth) {
        // Thanks Piyush
        int width = bm.getWidth();
        int height = bm.getHeight();
        float scaleWidth = ((float) newWidth) / width;
        float scaleHeight = ((float) newHeight) / height;
        // create a matrix for the manipulation
        Matrix matrix = new Matrix();
        // resize the bit map
        matrix.postScale(scaleWidth, scaleHeight);
        // recreate the new Bitmap
        return Bitmap.createBitmap(bm, 0, 0, width, height, matrix, false);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (enabled) {
            this.getParent().requestDisallowInterceptTouchEvent(true);

            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    if (onValuesChangeListener != null) {
                        onValuesChangeListener.onStartTrackingTouch(this);
                    }
                    if (!touchDisabled) updateOnTouch(event);
                    break;
                case MotionEvent.ACTION_MOVE:
                    updateOnTouch(event);
                    break;
                case MotionEvent.ACTION_UP:
                case MotionEvent.ACTION_CANCEL:
                    if (onValuesChangeListener != null) {
                        onValuesChangeListener.onStopTrackingTouch(this);
                    }
                    if (snapEnabled) setValue(getRoundedProgress());
                    setPressed(false);
                    this.getParent().requestDisallowInterceptTouchEvent(false);
                    break;
            }
            return true;
        }
        return false;
    }

    /**
     * Update the UI components on touch events.
     *
     * @param event MotionEvent
     */
    private void updateOnTouch(MotionEvent event) {
        setPressed(true);
        double mTouch = convertTouchEventPoint(event.getY());
        int progress = (int) Math.round(mTouch);
        updateProgress(progress);
    }

    private double convertTouchEventPoint(float yPos) {
        float wReturn;

        if (yPos > (scrHeight * 2)) {
            wReturn = scrHeight * 2;
            return wReturn;
        } else if (yPos < 0) {
            wReturn = 0;
        } else {
            wReturn = yPos;
        }

        return wReturn;
    }

    private void updateProgress(int progress) {
        progressSweep = progress;

        progress = Math.min(progress, scrHeight);
        progress = Math.max(progress, 0);

        // convert progress to min-max range
        points = progress * (max - min) / scrHeight + min;
        // reverse value because progress is descending
        points = max + min - points;
        // if value is not max or min, apply step
        if (points != max && points != min) {
            points = points - (points % step) + (min % step);
        }

        if (onValuesChangeListener != null) {
            onValuesChangeListener.onPointsChanged(this, points);
        }

        pointsText = String.valueOf(getRoundedProgress());

        invalidate();
    }

    private int getRoundedProgress() {
        return Math.round((scrHeight - progressSweep) / scrHeight * (max - min));
    }

    /**
     * Gets a value, converts it to progress for the seekBar and updates it.
     *
     * @param value The value given
     */
    private void updateProgressByValue(int value) {
        points = value;

        points = Math.min(points, max);
        points = Math.max(points, min);

        // convert min-max range to progress
        progressSweep = (points - min) * scrHeight / (max - min);
        // reverse value because progress is descending
        progressSweep = scrHeight - progressSweep;

        if (onValuesChangeListener != null) {
            onValuesChangeListener.onPointsChanged(this, points);
        }

        pointsText = String.valueOf(value);

        invalidate();
    }

    public interface OnValuesChangeListener {
        /**
         * Notification that the point value has changed.
         *
         * @param seekbar The Seekbar view whose value has changed
         * @param points  The current point value.
         */
        void onPointsChanged(BoxedVerticalSeekbar seekbar, int points);

        void onStartTrackingTouch(BoxedVerticalSeekbar seekbar);

        void onStopTrackingTouch(BoxedVerticalSeekbar seekbar);
    }

    public void setValue(int points) {
        points = Math.min(points, max);
        points = Math.max(points, min);

        updateProgressByValue(points);
    }

    public int getValue() {
        return points;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public int getMin() {
        return min;
    }

    public void setMin(int min) {
        this.min = min;
    }

    public int getMax() {
        return max;
    }

    public void setMax(int mMax) {
        if (mMax <= min) throw new IllegalArgumentException("Max should not be less than zero");
        this.max = mMax;
    }

    public void setCornerRadius(int mRadius) {
        this.cornerRadius = mRadius;
        invalidate();
    }

    public int getCornerRadius() {
        return cornerRadius;
    }

    public int getDefaultValue() {
        return defaultValue;
    }

    public void setDefaultValue(int mDefaultValue) {
        if (mDefaultValue > max)
            throw new IllegalArgumentException("Default value should not be bigger than max value.");
        this.defaultValue = mDefaultValue;
    }

    public int getStep() {
        return step;
    }

    public void setStep(int step) {
        this.step = step;
    }

    public boolean isImageEnabled() {
        return imageEnabled;
    }

    public void setImageEnabled(boolean mImageEnabled) {
        this.imageEnabled = mImageEnabled;
    }

    public boolean isSnapEnabled() {
        return snapEnabled;
    }

    public void setSnapEnabled(boolean snapEnabled) {
        this.snapEnabled = snapEnabled;
    }

    public void setOnBoxedPointsChangeListener(OnValuesChangeListener onValuesChangeListener) {
        this.onValuesChangeListener = onValuesChangeListener;
    }
}

File: core-android/src/main/java/com/lookaround/core/android/view/BoxedVerticalSeekbar.kt
package com.lookaround.core.android.view

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import androidx.core.content.ContextCompat
import com.lookaround.core.android.R
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

class BoxedVerticalSeekbar : View {
    /** The min value of progress value. */
    var min = 0

    /** The Maximum value that this SeekArc can be set to */
    var max = 100
        set(value) {
            require(value > min) { "Max should not be less than min value" }
            field = value
        }

    var defaultValue = 0
        set(value) {
            require(value <= max) { "Default value should not be bigger than max value." }
            field = value
        }

    /** The increment/decrement value for each movement of progress. */
    var step = 10

    /** The corner radius of the view. */
    var cornerRadius = 10
        set(value) {
            field = value
            invalidate()
        }

    /** Text size in SP. */
    private var textSize = 26f

    /** Text bottom padding in pixel. */
    private var textBottomPadding = 20
    private var points = 0
    private var enabled = true

    /** Enable or disable text . */
    private var textEnabled = true

    /** Enable or disable image . */
    var isImageEnabled = false
    var isSnapEnabled = true

    /** mTouchDisabled touches will not move the slider only swipe motion will activate it */
    private var touchDisabled = true
    private var progressSweep = 0f
    private val seekbarPaint = Paint()
    private val progressPaint: Paint = Paint()
    private val textPaint: Paint = Paint()
    private var scrWidth = 0
    private var scrHeight = 0
    var onValuesChangeListener: OnValuesChangeListener? = null
    private var defaultImage: Bitmap? = null
    private var minImage: Bitmap? = null
    private var maxImage: Bitmap? = null
    private val rect = Rect()
    private var firstRun = true
    private var pointsText = defaultValue.toString()

    constructor(context: Context) : super(context) {
        init(context, null)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        val density = resources.displayMetrics.density

        // Defaults, may need to link this into theme settings.
        var progressColor = ContextCompat.getColor(context, R.color.color_progress)
        var backgroundColor = ContextCompat.getColor(context, R.color.color_background)
        var textColor = ContextCompat.getColor(context, R.color.color_text)
        textSize *= density
        defaultValue = max / 2

        if (attrs != null) {
            val styledAttrs =
                context.obtainStyledAttributes(attrs, R.styleable.BoxedVerticalSeekbar, 0, 0)
            points = styledAttrs.getInteger(R.styleable.BoxedVerticalSeekbar_points, points)
            max = styledAttrs.getInteger(R.styleable.BoxedVerticalSeekbar_max, max)
            min = styledAttrs.getInteger(R.styleable.BoxedVerticalSeekbar_min, min)
            step = styledAttrs.getInteger(R.styleable.BoxedVerticalSeekbar_step, step)
            defaultValue =
                styledAttrs.getInteger(R.styleable.BoxedVerticalSeekbar_defaultValue, defaultValue)
            cornerRadius =
                styledAttrs.getInteger(
                    R.styleable.BoxedVerticalSeekbar_libCornerRadius,
                    cornerRadius
                )
            textBottomPadding =
                styledAttrs.getInteger(
                    R.styleable.BoxedVerticalSeekbar_textBottomPadding,
                    textBottomPadding
                )

            // Images
            isImageEnabled =
                styledAttrs.getBoolean(
                    R.styleable.BoxedVerticalSeekbar_imageEnabled,
                    isImageEnabled
                )
            if (isImageEnabled) {
                defaultImage =
                    requireNotNull(
                            styledAttrs.getDrawable(
                                R.styleable.BoxedVerticalSeekbar_defaultImage
                            ) as?
                                BitmapDrawable
                        ) {
                            "When images are enabled, defaultImage can not be null. Please assign a drawable in the layout XML file"
                        }
                        .bitmap
                minImage =
                    requireNotNull(
                            styledAttrs.getDrawable(R.styleable.BoxedVerticalSeekbar_minImage) as?
                                BitmapDrawable
                        ) {
                            "When images are enabled, minImage can not be null. Please assign a drawable in the layout XML file"
                        }
                        .bitmap
                maxImage =
                    requireNotNull(
                            styledAttrs.getDrawable(R.styleable.BoxedVerticalSeekbar_maxImage) as?
                                BitmapDrawable
                        ) {
                            "When images are enabled, maxImage can not be null. Please assign a drawable in the layout XML file"
                        }
                        .bitmap
            }

            progressColor =
                styledAttrs.getColor(R.styleable.BoxedVerticalSeekbar_progressColor, progressColor)
            backgroundColor =
                styledAttrs.getColor(
                    R.styleable.BoxedVerticalSeekbar_backgroundColor,
                    backgroundColor
                )
            textSize = styledAttrs.getDimension(R.styleable.BoxedVerticalSeekbar_textSize, textSize)
            textColor = styledAttrs.getColor(R.styleable.BoxedVerticalSeekbar_textColor, textColor)
            enabled = styledAttrs.getBoolean(R.styleable.BoxedVerticalSeekbar_enabled, enabled)
            touchDisabled =
                styledAttrs.getBoolean(
                    R.styleable.BoxedVerticalSeekbar_touchDisabled,
                    touchDisabled
                )
            textEnabled =
                styledAttrs.getBoolean(R.styleable.BoxedVerticalSeekbar_textEnabled, textEnabled)
            isSnapEnabled =
                styledAttrs.getBoolean(R.styleable.BoxedVerticalSeekbar_snapEnabled, isSnapEnabled)
            points = defaultValue
            styledAttrs.recycle()
        }

        // range check
        points = min(points, max)
        points = max(points, min)

        progressPaint.color = progressColor
        progressPaint.isAntiAlias = true
        progressPaint.style = Paint.Style.STROKE

        textPaint.color = textColor
        textPaint.isAntiAlias = true
        textPaint.style = Paint.Style.FILL
        textPaint.textSize = textSize

        seekbarPaint.alpha = 255
        seekbarPaint.color = backgroundColor
        seekbarPaint.isAntiAlias = true

        scrHeight = context.resources.displayMetrics.heightPixels
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        scrWidth = getDefaultSize(suggestedMinimumWidth, widthMeasureSpec)
        scrHeight = getDefaultSize(suggestedMinimumHeight, heightMeasureSpec)
        progressPaint.strokeWidth = scrWidth.toFloat()
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    override fun onDraw(canvas: Canvas) {
        canvas.translate(0f, 0f)
        val path = Path()
        path.addRoundRect(
            RectF(0f, 0f, scrWidth.toFloat(), scrHeight.toFloat()),
            cornerRadius.toFloat(),
            cornerRadius.toFloat(),
            Path.Direction.CCW
        )
        canvas.clipPath(path, Region.Op.INTERSECT)
        canvas.drawRect(0f, 0f, scrWidth.toFloat(), scrHeight.toFloat(), seekbarPaint)
        canvas.drawLine(
            width.toFloat() / 2,
            height.toFloat(),
            width.toFloat() / 2,
            progressSweep,
            progressPaint
        )
        if (isImageEnabled) {
            // If image is enabled, text will not be shown
            when (points) {
                max -> drawIcon(requireNotNull(maxImage), canvas)
                min -> drawIcon(requireNotNull(minImage), canvas)
                else -> drawIcon(requireNotNull(defaultImage), canvas)
            }
        } else {
            // If image is disabled and text is enabled show text
            if (textEnabled) {
                drawText(canvas, textPaint, pointsText)
            }
        }
        if (firstRun) {
            firstRun = false
            value = points
        }
    }

    private fun drawText(canvas: Canvas, paint: Paint, text: String) {
        canvas.getClipBounds(rect)
        val cWidth = rect.width()
        paint.textAlign = Paint.Align.LEFT
        paint.getTextBounds(text, 0, text.length, rect)
        val x = cWidth / 2f - rect.width() / 2f - rect.left
        canvas.drawText(text, x, (canvas.height - textBottomPadding).toFloat(), paint)
    }

    private fun drawIcon(bitmap: Bitmap, canvas: Canvas) {
        val resizedBitmap = getResizedBitmap(bitmap, canvas.width / 2, canvas.width / 2)
        canvas.drawBitmap(
            resizedBitmap,
            null,
            RectF(
                canvas.width.toFloat() / 2 - resizedBitmap.width.toFloat() / 2,
                (canvas.height - resizedBitmap.height).toFloat(),
                canvas.width.toFloat() / 3 + resizedBitmap.width,
                canvas.height.toFloat()
            ),
            null
        )
    }

    private fun getResizedBitmap(bm: Bitmap, newHeight: Int, newWidth: Int): Bitmap {
        val width = bm.width
        val height = bm.height
        val scaleWidth = newWidth.toFloat() / width
        val scaleHeight = newHeight.toFloat() / height
        // create a matrix for the manipulation
        val matrix = Matrix()
        // resize the bit map
        matrix.postScale(scaleWidth, scaleHeight)
        // recreate the new Bitmap
        return Bitmap.createBitmap(bm, 0, 0, width, height, matrix, false)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (enabled) {
            this.parent.requestDisallowInterceptTouchEvent(true)
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    onValuesChangeListener?.onStartTrackingTouch(this)
                    if (!touchDisabled) updateOnTouch(event)
                }
                MotionEvent.ACTION_MOVE -> updateOnTouch(event)
                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                    onValuesChangeListener?.onStopTrackingTouch(this)
                    if (isSnapEnabled) value = roundedProgress
                    isPressed = false
                    this.parent.requestDisallowInterceptTouchEvent(false)
                }
            }
            return true
        }
        return false
    }

    /**
     * Update the UI components on touch events.
     *
     * @param event MotionEvent
     */
    private fun updateOnTouch(event: MotionEvent) {
        isPressed = true
        val touchPoint = convertTouchEventPoint(event.y)
        val progress = touchPoint.roundToInt()
        updateProgress(progress)
    }

    private fun convertTouchEventPoint(yPos: Float): Double =
        when {
            yPos > scrHeight * 2 -> (scrHeight * 2).toFloat()
            yPos < 0 -> 0f
            else -> yPos
        }.toDouble()

    private fun updateProgress(progress: Int) {
        progressSweep = progress.toFloat()
        val coalescedProgress = max(min(progress, scrHeight), 0)

        // convert progress to min-max range
        points = coalescedProgress * (max - min) / scrHeight + min
        // reverse value because progress is descending
        points = max + min - points
        // if value is not max or min, apply step
        if (points != max && points != min) {
            points = points - points % step + min % step
        }
        onValuesChangeListener?.onPointsChanged(this, points)
        pointsText = roundedProgress.toString()
        invalidate()
    }

    private val roundedProgress: Int
        get() = ((scrHeight - progressSweep) / scrHeight * (max - min)).roundToInt()

    /**
     * Gets a value, converts it to progress for the seekBar and updates it.
     *
     * @param value The value given
     */
    private fun updateProgressByValue(value: Int) {
        points = value
        points = min(points, max)
        points = max(points, min)

        // convert min-max range to progress
        progressSweep = ((points - min) * scrHeight / (max - min)).toFloat()
        // reverse value because progress is descending
        progressSweep = scrHeight - progressSweep
        onValuesChangeListener?.onPointsChanged(this, points)
        pointsText = value.toString()
        invalidate()
    }

    interface OnValuesChangeListener {
        fun onPointsChanged(seekbar: BoxedVerticalSeekbar, points: Int) = Unit
        fun onStartTrackingTouch(seekbar: BoxedVerticalSeekbar) = Unit
        fun onStopTrackingTouch(seekbar: BoxedVerticalSeekbar) = Unit
    }

    var value: Int
        get() = points
        set(points) {
            updateProgressByValue(max(min(points, max), min))
        }

    override fun isEnabled(): Boolean = enabled

    override fun setEnabled(enabled: Boolean) {
        this.enabled = enabled
    }
}
